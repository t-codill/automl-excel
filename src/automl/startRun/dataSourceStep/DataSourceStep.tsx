import { AzureMachineLearningWorkspacesModels } from "@azure/arm-machinelearningservices";
import { StorageManagementModels } from "@azure/arm-storage";
import { Models } from "@azure/storage-blob";
import { TextField } from "office-ui-fabric-react";
import * as React from "react";
import { ICsvData } from "../../common/utils/csv";
import { IDefaultStorageAccountData } from "../StartRun";
import { BlobPreviewer } from "./BlobPreviewer";
import { BlobSelector } from "./BlobSelector";
import { StorageAccountSelector } from "./StorageAccountSelector";
import { StorageContainerSelector } from "./StorageContainerSelector";

export interface IDataSourceStepProps {
    account: StorageManagementModels.StorageAccount | undefined;
    sasToken: string | undefined;
    container: StorageManagementModels.ListContainerItem | undefined;
    blob: Models.BlobItem | undefined;
    previewData: ICsvData | undefined;
    dataStoreName: string | undefined;
    compute: AzureMachineLearningWorkspacesModels.ComputeResource;
    defaultStorageAccountData: IDefaultStorageAccountData | undefined;
    onDataSourceChanged(
        account: StorageManagementModels.StorageAccount | undefined,
        sasToken: string | undefined,
        container: StorageManagementModels.ListContainerItem | undefined,
        blob: Models.BlobItem | undefined,
        previewData: ICsvData | undefined
    ): void;
    onFeatureSelectionChanged(featureName: string, checked?: boolean): void;
}
export interface IDataSourceStepState {
    step: string;
    autoSelect: boolean;
}
export class DataSourceStep
    extends React.Component<IDataSourceStepProps, IDataSourceStepState> {

    constructor(props: IDataSourceStepProps) {
        super(props);

        this.state = {
            step: "Account",
            autoSelect: true
        };
    }

    public render(): React.ReactNode {
        return <>
            <div className="start-run-step-content">
                {this.renderDataSourceAccount()}
                {this.renderDataSourceContainer()}
                {this.renderDataSourceBlob()}
                {this.renderDataSourcePreview()}
            </div>
        </>;
    }

    private readonly renderDataSourceAccount = (): React.ReactNode => {
        if (this.props.account && this.props.sasToken) {
            return <div className="ms-md12 ms-lg8 ms-xl6 ms-xxl4">
                <TextField
                    label="Storage Account"
                    ariaLabel="Storage Account"
                    readOnly={true}
                    iconProps={
                        {
                            iconName: "Edit",
                            style: { fontSize: 12 }
                        }
                    }
                    value={this.props.account.name}
                    onClick={this.changeAccount}
                />
            </div>;
        }
        return <StorageAccountSelector
            account={this.props.account}
            onAccountSelected={this.onAccountSelected}
            defaultStorageAccountName={this.props.defaultStorageAccountData && this.props.defaultStorageAccountData.defaultAccountName}
            autoSelect={this.state.autoSelect}
        />;
    }

    private readonly changeAccount = () => {
        this.setState({ autoSelect: false });
        this.props.onDataSourceChanged(undefined, undefined, undefined, undefined, undefined);
    }

    private readonly onAccountSelected = (
        account: StorageManagementModels.StorageAccount,
        sasToken: string) => {
        this.props.onDataSourceChanged(account, sasToken, undefined, undefined, undefined);
    }

    private readonly renderDataSourceContainer = (): React.ReactNode => {
        if (!this.props.account || !this.props.sasToken) {
            return undefined;
        }
        if (this.props.container) {
            return <div className="ms-md12 ms-lg8 ms-xl6 ms-xxl4">
                <TextField
                    label="Storage Container"
                    ariaLabel="Storage Container"
                    readOnly={true}
                    iconProps={
                        {
                            iconName: "Edit",
                            style: { fontSize: 12 }
                        }
                    }
                    value={this.props.container.name}
                    onClick={this.changeContainer}
                />
            </div>;
        }
        return <StorageContainerSelector
            account={this.props.account}
            container={this.props.container}
            defaultContainerName={this.props.defaultStorageAccountData && this.props.defaultStorageAccountData.defaultContainerName}
            onContainerSelected={this.onContainerSelected}
            autoSelect={this.state.autoSelect} />;
    }

    private readonly changeContainer = () => {
        this.props.onDataSourceChanged(this.props.account, this.props.sasToken, undefined, undefined, undefined);
    }

    private readonly onContainerSelected = (container: StorageManagementModels.ListContainerItem) => {
        this.setState({ autoSelect: false });
        this.props.onDataSourceChanged(this.props.account, this.props.sasToken, container, undefined, undefined);
    }

    private readonly renderDataSourceBlob = (): React.ReactNode => {
        if (!this.props.account
            || !this.props.sasToken
            || !this.props.container) {
            return undefined;
        }
        if (this.props.blob) {
            return <div className="ms-md12 ms-lg8 ms-xl6 ms-xxl4">
                <TextField
                    label="Blob"
                    ariaLabel="Blob"
                    readOnly={true}
                    iconProps={
                        {
                            iconName: "Edit",
                            style: { fontSize: 12 }
                        }
                    }
                    value={this.props.blob.name}
                    onClick={this.changeBlob}
                />
            </div>;
        }
        return <BlobSelector
            account={this.props.account}
            container={this.props.container}
            sasToken={this.props.sasToken}
            blob={this.props.blob}
            onBlobSelected={this.onBlobSelected} />;
    }

    private readonly changeBlob = () => {
        this.props.onDataSourceChanged(this.props.account, this.props.sasToken, this.props.container, undefined, undefined);
    }

    private readonly onBlobSelected = (blob: Models.BlobItem) => {
        this.props.onDataSourceChanged(this.props.account, this.props.sasToken, this.props.container, blob, undefined);
    }

    private readonly renderDataSourcePreview = (): React.ReactNode => {
        if (!this.props.account
            || !this.props.sasToken
            || !this.props.container
            || !this.props.blob) {
            return undefined;
        }
        return <BlobPreviewer
            account={this.props.account}
            container={this.props.container}
            sasToken={this.props.sasToken}
            blob={this.props.blob}
            compute={this.props.compute}
            dataStoreName={this.props.dataStoreName}
            previewData={this.props.previewData}
            onSetPreviewData={this.onSetPreviewData}
            onFeatureSelectionChanged={this.props.onFeatureSelectionChanged} />;
    }

    private readonly onSetPreviewData = (previewData: ICsvData) => {
        this.props.onDataSourceChanged(this.props.account, this.props.sasToken, this.props.container, this.props.blob, previewData);
    }
}
