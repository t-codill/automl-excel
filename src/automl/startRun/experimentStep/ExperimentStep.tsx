import { AzureMachineLearningWorkspacesModels } from "@azure/arm-machinelearningservices";
import { map, uniq } from "lodash";
import { DefaultButton, Dialog, DialogFooter, DialogType, Icon, Link, PrimaryButton } from "office-ui-fabric-react";
import * as React from "react";
import { BaseComponent } from "../../components/Base/BaseComponent";
import { Form } from "../../components/Form/Form";
import { RunHistoryService } from "../../services/RunHistoryService";
import { IVMSizeResult, WorkSpaceService } from "../../services/WorkSpaceService";
import { ComputeCreator } from "./ComputeCreator";
import { ComputeSelector } from "./ComputeSelector";
import { ExperimentNameInput } from "./ExperimentNameInput";

export interface IExperimentStepProps {
    experimentName: string | undefined;
    compute: AzureMachineLearningWorkspacesModels.ComputeResource | undefined;
    readOnly: boolean;
    onEditClick(): void;
    onCancel(): void;
    onNext(
        experimentName: string | undefined,
        compute: AzureMachineLearningWorkspacesModels.ComputeResource | undefined
    ): void;
}
export interface IExperimentStepCallback {
    experimentName: string | undefined;
    computeId: string | undefined;
}

export interface IExperimentStepState {
    experimentNames: string[] | undefined;
    computes: AzureMachineLearningWorkspacesModels.ComputeResource[] | undefined;
    selectedCompute: AzureMachineLearningWorkspacesModels.ComputeResource | undefined;
    showComputeCreator: boolean;
    vmSizes: IVMSizeResult | undefined;
    showCancelDialog: boolean;
    computeRefreshing: boolean;
}

export class ExperimentStep
    extends BaseComponent<IExperimentStepProps, IExperimentStepState, {
        workSpaceService: WorkSpaceService;
        runHistoryService: RunHistoryService;
    }> {
    protected readonly serviceConstructors = {
        workSpaceService: WorkSpaceService,
        runHistoryService: RunHistoryService
    };

    constructor(props: IExperimentStepProps) {
        super(props);
        this.state = {
            experimentNames: undefined,
            computes: undefined,
            selectedCompute: props.compute,
            showComputeCreator: false,
            vmSizes: undefined,
            showCancelDialog: false,
            computeRefreshing: false
        };
    }

    public render(): React.ReactNode {
        return <>
            <Form
                onSubmit={this.onSubmit}>
                <div className="start-run-step-content">
                    <div className="ms-md12 ms-lg8 ms-xl6 ms-xxl4">
                        <ExperimentNameInput
                            experimentName={this.props.experimentName}
                            readonly={this.props.readOnly}
                            onEditClick={this.props.onEditClick}
                            experimentNames={this.state.experimentNames}
                        />
                    </div>
                    <div className="ms-md12 ms-lg8 ms-xl6 ms-xxl4">
                        <ComputeSelector
                            computes={this.state.computes}
                            selectedComputeId={this.state.selectedCompute && this.state.selectedCompute.id}
                            readonly={this.props.readOnly}
                            refreshing={this.state.computeRefreshing}
                            onEditClick={this.props.onEditClick}
                        />
                        {!this.props.readOnly &&
                            <Link
                                id="linkCreateCompute"
                                onClick={this.onCreateComputeClick}>
                                <Icon iconName="Devices3" style={{ marginRight: 5 }} />
                                <span>Create a new compute</span>
                            </Link>
                        }
                        {!this.props.readOnly &&
                            <Link
                                id="linkRefreshCompute"
                                onClick={this.onRefreshComputeClick} >
                                <Icon iconName="Refresh" style={{ marginRight: 5, marginLeft: 15 }} />
                                <span>Refresh Compute</span>
                            </Link>
                        }
                    </div>
                </div>

                {!this.props.readOnly &&
                    <div className="form-footer">
                        <DefaultButton text="Cancel" onClick={this.showCancelDialog} />
                        <PrimaryButton type="submit" text="Next" />
                    </div>
                }
            </Form>

            {this.state.showComputeCreator &&
                <ComputeCreator
                    {...this.props}
                    {...this.state}
                    onComputeCreated={this.onComputeCreated}
                    onCancel={this.onComputeCreateCancel}
                />
            }
            <Dialog
                hidden={!this.state.showCancelDialog}
                onDismiss={this.closeCancelDialog}
                dialogContentProps={{
                    type: DialogType.normal,
                    title: "Cancel New Experiment?",
                    subText: `Are you sure you want to cancel creating a new Experiment?`
                }}
                modalProps={{
                    titleAriaId: "cancelDialogLabel",
                    subtitleAriaId: "cancelSubtextLabel",
                    isBlocking: false,
                    containerClassName: "ms-dialogMainOverride"
                }}
            >
                <DialogFooter>
                    <PrimaryButton onClick={this.cancelRun} text="Yes" />
                    <DefaultButton onClick={this.closeCancelDialog} text="No" />
                </DialogFooter>
            </Dialog>
        </>;
    }

    protected readonly getData = async () => {
        this.getExperimentNames();
        this.getComputes();
        this.getVmSizes();
    }

    private readonly onCreateComputeClick = () => {
        this.setState({ showComputeCreator: true });
    }

    private readonly onRefreshComputeClick = () => {
        this.getComputes();
    }

    private readonly onComputeCreated = (compute: AzureMachineLearningWorkspacesModels.ComputeResource) => {
        this.setState({ showComputeCreator: false, selectedCompute: compute });
        this.getComputes();
    }

    private readonly onComputeCreateCancel = () => {
        this.setState({ showComputeCreator: false });
    }

    private readonly getExperimentNames = async (): Promise<void> => {
        const experiments = await this.services.runHistoryService.listExperiments();
        if (!experiments) {
            return;
        }
        const experimentNames = uniq(map(experiments, (experiment) => experiment.name || ""));
        this.setState({ experimentNames });
    }

    private readonly getComputes = async (): Promise<void> => {
        this.setState({ computeRefreshing: true });
        const computes = await this.services.workSpaceService.listComputes();
        if (!computes) {
            return;
        }
        this.setState({ computes, computeRefreshing: false });
    }

    private readonly getVmSizes = async (): Promise<void> => {
        const vmSizes = await this.services.workSpaceService.listVmSizes();
        if (!vmSizes) {
            return;
        }
        this.setState({ vmSizes });
    }

    private readonly onSubmit = (values: IExperimentStepCallback) => {
        if (!values.computeId || !this.state.computes) {
            return;
        }
        this.props.onNext(values.experimentName, this.state.computes.find((c) => c.id === values.computeId));
    }

    private readonly showCancelDialog = (): void => {
        this.setState({ showCancelDialog: true });
    }

    private readonly closeCancelDialog = (): void => {
        this.setState({ showCancelDialog: false });
    }

    private readonly cancelRun = () => {
        this.props.onCancel();
        this.closeCancelDialog();
    }

}
