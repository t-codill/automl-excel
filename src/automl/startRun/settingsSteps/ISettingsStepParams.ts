import { AzureMachineLearningJasmineManagementModels } from "@vienna/jasmine";
import { RunType } from "../../services/constants/RunType";

export interface ISettingsStepParams {
    jobType: RunType | undefined;
    column: string | undefined;
    metric: AzureMachineLearningJasmineManagementModels.PrimaryMetric | undefined;
    trainingJobTime: string | undefined;
    maxIteration: string | undefined;
    metricThreshold: string | undefined;
    maxConcurrentIterations: string | undefined;
    maxCores: string | undefined;
    preprocessing: boolean | undefined;
    blacklistAlgos: string[];
    crossValidationNumber: string | undefined;
    percentageValidation: string | undefined;
    timeSeriesColumn: string | undefined;
    maxHorizon: string | undefined;
    grainColumns: string[];
}
