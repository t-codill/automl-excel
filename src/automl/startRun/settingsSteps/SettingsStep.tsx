import { AzureMachineLearningWorkspacesModels } from "@azure/arm-machinelearningservices";
import { DefaultButton, PrimaryButton } from "office-ui-fabric-react";
import * as React from "react";
import { Accordion } from "../../components/Accordion/Accordion";
import { Form } from "../../components/Form/Form";
import { MetricType } from "../../services/constants/MetricType";
import { RunType } from "../../services/constants/RunType";
import { AdvancedSettings } from "./AdvancedSettings";
import { GrainColumnSelector } from "./GrainColumnSelector";
import { ISettingsStepParams } from "./ISettingsStepParams";
import { JobTypeSelector } from "./JobTypeSelector";
import { MaxHorizon } from "./MaxHorizon";
import { TargetColumnSelector } from "./TargetColumnSelector";
import { TimeSeriesColumnSelector } from "./TimeSeriesColumnSelector";

export interface ISettingsStepProps {
    blobHeader: string[];
    dataStoreName: string | undefined;
    compute: AzureMachineLearningWorkspacesModels.ComputeUnion | undefined;
    selectedFeatures: Set<string>;
    onCancel(): void;
    onStart(
        parameters: ISettingsStepParams
    ): void;
}

export interface ISettingsStepState {
    jobType: RunType;
    primaryMetric: string;
    column: string | undefined;
    timeSeriesColumn: string | undefined;
}

export class SettingsStep extends React.Component<ISettingsStepProps, ISettingsStepState> {
    constructor(props: ISettingsStepProps) {
        super(props);
        this.state = {
            jobType: "classification",
            primaryMetric: MetricType.classification[0],
            column: this.props.blobHeader[0],
            timeSeriesColumn: undefined
        };
    }

    public render(): React.ReactNode {
        return <Form onSubmit={this.onSubmit} onUpdated={this.onFormUpdated}>
            <div className="start-run-step-content">
                <div className="ms-md12 ms-lg8 ms-xl6 ms-xxl4">
                    <JobTypeSelector
                        jobType={this.state.jobType} />
                </div>
                <div className="ms-md12 ms-lg8 ms-xl6 ms-xxl4">
                    <TargetColumnSelector
                        columns={this.props.blobHeader} />
                </div>
                {
                    this.state.jobType === "forecasting" &&
                    <>
                        <div className="ms-md12 ms-lg8 ms-xl6 ms-xxl4">
                            <TimeSeriesColumnSelector
                                targetColumn={this.state.column}
                                selectedFeatures={this.props.selectedFeatures}
                                columns={this.props.blobHeader} />
                        </div>
                        <div className="ms-md12 ms-lg8 ms-xl6 ms-xxl4">
                            <GrainColumnSelector
                                targetColumn={this.state.column}
                                timeSeriesColumn={this.state.timeSeriesColumn}
                                selectedFeatures={this.props.selectedFeatures}
                                columns={this.props.blobHeader} />
                        </div>
                        <div className="ms-md12 ms-lg8 ms-xl6 ms-xxl4">
                            <MaxHorizon />
                        </div>
                    </>
                }
            </div>
            <Accordion
                exclusive={true}
                items={[
                    {
                        title: "Advanced Settings",
                        collapsed: true,
                        disabled: false,
                        element: <AdvancedSettings compute={this.props.compute}
                            jobType={this.state.jobType}
                            primaryMetric={this.state.primaryMetric} />
                    }
                ]} />
            <div className="form-footer">
                <DefaultButton text="Cancel" onClick={this.props.onCancel} />
                <PrimaryButton type="submit" text="Start" disabled={!this.props.dataStoreName} />
            </div>
        </Form>;
    }

    private readonly onFormUpdated = (key: keyof ISettingsStepParams, value: ISettingsStepParams[keyof ISettingsStepParams]) => {
        if (!value) {
            return;
        }
        if (key === "metric") {
            const metric = value.toString();
            if (metric !== this.state.primaryMetric) {
                this.setState({ primaryMetric: metric });
            }
        } else if (key === "column") {
            const column = value.toString();
            this.setState({ column });
        } else if (key === "timeSeriesColumn") {
            const timeSeriesColumn = value.toString();
            this.setState({ timeSeriesColumn });
        } else if (key === "jobType" && this.state.jobType !== value) {
            this.setState({ jobType: value as RunType });
        }
    }

    private readonly onSubmit = (values: ISettingsStepParams) => {
        this.props.onStart(values);
    }
}
