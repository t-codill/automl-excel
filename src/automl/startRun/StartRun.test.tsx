import { AzureMachineLearningWorkspacesModels } from "@azure/arm-machinelearningservices";
import { StorageManagementModels } from "@azure/arm-storage";
import { Models } from "@azure/storage-blob";
import { shallow, ShallowWrapper } from "enzyme";
import * as React from "react";
import { sampleStorageAccount } from "../../__data__/sampleStorageAccount";
import { sampleStorageBlob } from "../../__data__/sampleStorageBlob";
import { sampleStorageContainer } from "../../__data__/sampleStorageContainer";
import { getLogCustomEventSpy, testContext } from "../common/context/__data__/testContext";
import { ICsvData } from "../common/utils/csv";
import { IBasePageState } from "../components/Base/BasePage";
import { ConfirmationDialog } from "../components/Dialog/ConfirmationDialog";
import { PopupProgressIndicator } from "../components/Progress/PopupProgressIndicator";
import { sampleDataStore } from "../services/__data__/sampleDataStore";
import { DataStoreService } from "../services/DataStoreService";
import { JasmineService } from "../services/JasmineService";
import { DataSourceStep } from "./dataSourceStep/DataSourceStep";
import { ExperimentStep } from "./experimentStep/ExperimentStep";
import { ISettingsStepParams } from "./settingsSteps/ISettingsStepParams";
import { SettingsStep } from "./settingsSteps/SettingsStep";
import { IStartRunState, StartRun } from "./StartRun";

jest.mock("../services/DataStoreService");
jest.mock("../services/JasmineService");

describe("StartRun", () => {
    describe("should change status", () => {
        let tree: ShallowWrapper<{}, IStartRunState, StartRun>;
        beforeEach(async () => {
            tree = shallow(<StartRun />);
            await Promise.resolve();
        });

        it("should render", () => {
            expect(tree)
                .toMatchSnapshot();
        });

        it("should cancel", () => {
            tree.setState({ cancelled: true });
            expect(tree)
                .toMatchInlineSnapshot(`
      <PageRedirectRender
        expendedRoutePath=""
        noPush={false}
      />
    `
                );
        });

        it("should redirect to parent run", () => {
            tree.setState({
                created: true,
                experimentName: "experimentName",
                runId: "runId"
            });
            expect(tree)
                .toMatchInlineSnapshot(`
      <PageRedirectRender
        expendedRoutePath="experiments/experimentName/parentrun/runId"
        noPush={false}
      />
    `
                );
        });

        it("should show PopupProgressIndicator", () => {
            tree.setState({
                creating: true
            });
            expect(tree.exists(PopupProgressIndicator))
                .toBe(true);
        });
    });
    describe("cancel", () => {
        let tree: ShallowWrapper<{}, IStartRunState & IBasePageState, StartRun>;
        beforeEach(async () => {
            tree = shallow(<StartRun />);
            tree.setState({ goBack: true });
            await Promise.resolve();
        });
        it("should show cancel dialog", () => {
            expect(tree.exists(ConfirmationDialog))
                .toBe(true);
        });
        it("should hide cancel", () => {
            tree.find(ConfirmationDialog)
                .prop("onClose")();
            expect(tree.state("goBack"))
                .toBe(false);
        });
        it("should cancel", () => {
            tree.find(ConfirmationDialog)
                .prop("onConfirm")();
            expect(tree.state("cancelled"))
                .toBe(true);
        });
    });

    describe("default store", () => {
        let tree: ShallowWrapper<{}, IStartRunState, StartRun>;
        let getDefaultDataStoreSpy: jest.SpyInstance<ReturnType<DataStoreService["getDefault"]>>;
        beforeEach(async () => {
            getDefaultDataStoreSpy = jest.spyOn(DataStoreService.prototype, "getDefault");
            tree = shallow(<StartRun />);
            await Promise.resolve();
        });
        it("should load default store", () => {
            expect(tree.state("defaultStorageAccountData"))
                .toEqual({
                    defaultAccountName: sampleDataStore.azureStorageSection.accountName,
                    defaultContainerName: sampleDataStore.azureStorageSection.containerName
                });
        });
        it("should load default store if get default is canceled", async () => {
            getDefaultDataStoreSpy.mockReturnValue(Promise.resolve(undefined));
            tree = shallow(<StartRun />);
            await Promise.resolve();
            expect(tree.state("defaultStorageAccountData"))
                .toBeUndefined();
        });
    });

    describe("data store", () => {
        let tree: ShallowWrapper<{}, IStartRunState, StartRun>;
        let addDataStoreSpy: jest.SpyInstance<ReturnType<DataStoreService["add"]>>;
        beforeEach(async () => {
            addDataStoreSpy = jest.spyOn(DataStoreService.prototype, "add");
            tree = shallow(<StartRun />);
            await Promise.resolve();
        });
        it("should add", async () => {
            tree.setState({
                account: sampleStorageAccount,
                container: sampleStorageContainer,
                sasToken: "sampleSasToken"
            });
            await Promise.resolve();
            expect(addDataStoreSpy)
                .toBeCalledWith(sampleStorageContainer, sampleStorageAccount, "sampleSasToken");
            expect(tree.state("dataStoreName"))
                .toBe("dataStoreName");
        });
        it("should not add is missing token", async () => {
            tree.setState({
                account: sampleStorageAccount,
                container: sampleStorageContainer,
                sasToken: undefined
            });
            await Promise.resolve();
            expect(addDataStoreSpy)
                .not
                .toBeCalled();
        });
        it("should not set data store name if add to store is canceled", async () => {
            addDataStoreSpy.mockReturnValue(Promise.resolve(undefined));
            tree.setState({
                account: sampleStorageAccount,
                container: sampleStorageContainer,
                sasToken: "sampleSasToken"
            });
            await Promise.resolve();
            expect(addDataStoreSpy)
                .toBeCalled();
            expect(tree.state("dataStoreName"))
                .toBeUndefined();
        });
    });

    describe("start run flow", () => {
        let tree: ShallowWrapper<{}, IStartRunState & IBasePageState, StartRun>;
        beforeAll(async () => {
            tree = shallow(<StartRun />);
            await Promise.resolve();
        });

        describe("Experiment Step", () => {
            let onExperimentStepNext: (experimentName: string | undefined,
                compute: AzureMachineLearningWorkspacesModels.ComputeResource | undefined) => void;
            let onEditExperimentStepClicked: () => void;
            const experimentName = "experimentName";
            const compute = { id: "sampleCompute" };

            beforeEach(() => {
                onExperimentStepNext = tree.find(ExperimentStep)
                    .prop("onNext");
                onEditExperimentStepClicked = tree.find(ExperimentStep)
                    .prop("onEditClick");
            });

            it("should go next", () => {
                onExperimentStepNext(experimentName, compute);
                expect(tree.state())
                    .toEqual(
                        expect.objectContaining({
                            experimentName,
                            compute,
                            experimentStepReadOnly: true
                        }));
            });

            it("should go back", () => {
                onEditExperimentStepClicked();
                expect(tree.state("experimentStepReadOnly"))
                    .toEqual(false);
                onExperimentStepNext(experimentName, compute);
            });
        });

        describe("DataSource Step", () => {
            let onDataSourceChanged: (
                account: StorageManagementModels.StorageAccount | undefined,
                sasToken: string | undefined,
                container: StorageManagementModels.ListContainerItem | undefined,
                blob: Models.BlobItem | undefined,
                previewData: ICsvData | undefined) => void;
            let onFeatureSelectionChanged: (featureName: string, checked?: boolean) => void;
            const previewData: ICsvData = {
                data: [
                    { a: "1-1", b: "1-2", c: "1-3" },
                    { a: "2-1", b: "2-2", c: "2-3" },
                    { a: "3-1", b: "3-2", c: "3-3" }
                ],
                header: ["a", "b", "c"],
                delimiter: ",",
                hasHeader: true
            };

            beforeEach(() => {
                onDataSourceChanged = tree.find(DataSourceStep)
                    .prop("onDataSourceChanged");
                onFeatureSelectionChanged = tree.find(DataSourceStep)
                    .prop("onFeatureSelectionChanged");
            });
            it("should render", () => {
                expect(tree)
                    .toMatchSnapshot();
            });

            it("should have empty feature", () => {
                onDataSourceChanged(
                    undefined,
                    undefined,
                    undefined,
                    undefined,
                    undefined
                );
                expect(tree.state("selectedFeatures"))
                    .toEqual(new Set());
            });

            it("should onDataSourceChanged", () => {
                onDataSourceChanged(
                    sampleStorageAccount,
                    "sasToken",
                    sampleStorageContainer,
                    sampleStorageBlob,
                    previewData
                );
                expect(tree.state())
                    .toEqual(
                        expect.objectContaining({
                            account: sampleStorageAccount,
                            sasToken: "sasToken",
                            container: sampleStorageContainer,
                            blob: sampleStorageBlob,
                            previewData,
                            selectedFeatures: new Set(previewData.header)
                        }));
            });

            it("should deselect feature", () => {
                onFeatureSelectionChanged("a", false);
                expect(tree.state("selectedFeatures"))
                    .toEqual(new Set(["b", "c"]));
            });

            it("should select feature", () => {
                onFeatureSelectionChanged("a", true);
                expect(tree.state("selectedFeatures"))
                    .toEqual(new Set(["b", "c", "a"]));
            });

            it("should not select feature", () => {
                onFeatureSelectionChanged("a", undefined);
                expect(tree.state("selectedFeatures"))
                    .toEqual(new Set(["b", "c", "a"]));
            });

        });

        describe("Settings Step", () => {
            let onStart: (params: ISettingsStepParams) => void;
            let onCancel: () => void;
            let logger: jest.SpyInstance;
            let createRunSpy: jest.SpyInstance<ReturnType<JasmineService["createRun"]>>;
            let startRunSpy: jest.SpyInstance<ReturnType<JasmineService["startRun"]>>;
            const param: ISettingsStepParams = {
                blacklistAlgos: [],
                jobType: "classification",
                column: "a",
                metric: "AUC_weighted",
                trainingJobTime: "90",
                maxIteration: "10",
                crossValidationNumber: "2",
                grainColumns: [],
                maxConcurrentIterations: "2",
                maxCores: "1",
                maxHorizon: undefined,
                metricThreshold: undefined,
                percentageValidation: undefined,
                preprocessing: true,
                timeSeriesColumn: undefined
            };

            beforeEach(() => {
                createRunSpy = jest.spyOn(JasmineService.prototype, "createRun");
                startRunSpy = jest.spyOn(JasmineService.prototype, "startRun");
                onStart = tree.find(SettingsStep)
                    .prop("onStart");
                onCancel = tree.find(SettingsStep)
                    .prop("onCancel");
                logger = getLogCustomEventSpy();
            });
            it("should render", () => {
                expect(tree)
                    .toMatchSnapshot();
            });

            it("should cancel", () => {
                onCancel();
                expect(tree.state("goBack"))
                    .toBe(true);
                tree.setState({ cancelled: false });
            });

            it("should hide if user go back to experiment step", () => {
                tree.setState({ experimentStepReadOnly: false });
                expect(tree.find("#divDataSourceStep")
                    .hasClass("hidden_step"))
                    .toBe(true);
                tree.setState({ experimentStepReadOnly: true });
            });

            it("should not continue without experiment name", async () => {
                tree.setState({ experimentName: undefined });
                onStart(param);
                expect(tree.state("creating"))
                    .toBe(false);
            });

            it("should not continue without data store name", async () => {
                tree.setState({ experimentName: "experimentName", dataStoreName: undefined });
                onStart(param);
                expect(tree.state("creating"))
                    .toBe(true);
                expect(tree.state("creatingDescription"))
                    .toBe("Creating data store...");
            });

            it("create run missing param", async () => {
                tree.setState({ dataStoreName: "dataStoreName" });
                onStart({ ...param, jobType: undefined });
                await Promise.resolve();
                expect(tree.state("creating"))
                    .toBe(false);
                expect(tree.state("created"))
                    .toBe(false);
            });

            it("create run error", async () => {
                tree.setState({ dataStoreName: "dataStoreName" });
                createRunSpy.mockReturnValue(Promise.resolve(undefined));
                onStart(param);
                await Promise.resolve();
                await Promise.resolve();
                await Promise.resolve();
                expect(logger)
                    .toBeCalledWith("StartRun_CreateRun_UserAction", testContext, expect.anything());
                expect(tree.state("creating"))
                    .toBe(false);
                expect(tree.state("created"))
                    .toBe(false);
            });

            it("create run, and start run canceled", async () => {
                startRunSpy.mockReturnValue(Promise.resolve(undefined));
                onStart(param);
                await Promise.resolve();
                await Promise.resolve();
                await Promise.resolve();
                await Promise.resolve();
                await Promise.resolve();
                await Promise.resolve();
                expect(tree.state("runId"))
                    .toBe("AutoML_123");
                expect(tree.state("creating"))
                    .toBe(false);
                expect(tree.state("created"))
                    .toBe(false);
            });

            it("started", async () => {
                onStart(param);
                await Promise.resolve();
                await Promise.resolve();
                await Promise.resolve();
                await Promise.resolve();
                await Promise.resolve();
                await Promise.resolve();
                expect(tree.state("created"))
                    .toBe(true);
            });

        });

    });

});
