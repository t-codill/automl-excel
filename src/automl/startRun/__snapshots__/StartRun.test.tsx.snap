// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StartRun should change status should render 1`] = `
<Fragment>
  <ExperimentStep
    onCancel={[Function]}
    onEditClick={[Function]}
    onNext={[Function]}
    readOnly={false}
  />
</Fragment>
`;

exports[`StartRun start run flow DataSource Step should render 1`] = `
<Fragment>
  <ExperimentStep
    compute={
      Object {
        "id": "sampleCompute",
      }
    }
    experimentName="experimentName"
    onCancel={[Function]}
    onEditClick={[Function]}
    onNext={[Function]}
    readOnly={true}
  />
  <div
    id="divDataSourceStep"
  >
    <DataSourceStep
      compute={
        Object {
          "id": "sampleCompute",
        }
      }
      defaultStorageAccountData={
        Object {
          "defaultAccountName": "account",
          "defaultContainerName": "defaultContainer",
        }
      }
      onDataSourceChanged={[Function]}
      onFeatureSelectionChanged={[Function]}
    />
  </div>
</Fragment>
`;

exports[`StartRun start run flow Settings Step should render 1`] = `
<Fragment>
  <ExperimentStep
    compute={
      Object {
        "id": "sampleCompute",
      }
    }
    experimentName="experimentName"
    onCancel={[Function]}
    onEditClick={[Function]}
    onNext={[Function]}
    readOnly={true}
  />
  <div
    id="divDataSourceStep"
  >
    <DataSourceStep
      account={
        Object {
          "id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/sampleResourceGroup/Microsoft.Storage/storageAccounts/sampleStorageAccounts",
          "location": "eastus",
          "name": "sampleStorageAccount",
          "primaryEndpoints": Object {
            "blob": "https://blob.samplestorageaccount.com",
          },
        }
      }
      blob={
        Object {
          "deleted": false,
          "name": "sampleBlob",
          "properties": Object {
            "contentLength": 1000,
            "etag": "sampleEtag",
            "lastModified": 2019-01-01T00:00:00.000Z,
          },
          "snapshot": "sampleSnapShot",
        }
      }
      compute={
        Object {
          "id": "sampleCompute",
        }
      }
      container={
        Object {
          "name": "sampleContainer",
        }
      }
      dataStoreName="dataStoreName"
      defaultStorageAccountData={
        Object {
          "defaultAccountName": "account",
          "defaultContainerName": "defaultContainer",
        }
      }
      onDataSourceChanged={[Function]}
      onFeatureSelectionChanged={[Function]}
      previewData={
        Object {
          "data": Array [
            Object {
              "a": "1-1",
              "b": "1-2",
              "c": "1-3",
            },
            Object {
              "a": "2-1",
              "b": "2-2",
              "c": "2-3",
            },
            Object {
              "a": "3-1",
              "b": "3-2",
              "c": "3-3",
            },
          ],
          "delimiter": ",",
          "hasHeader": true,
          "header": Array [
            "a",
            "b",
            "c",
          ],
        }
      }
      sasToken="sasToken"
    />
  </div>
  <div>
    <SettingsStep
      blobHeader={
        Array [
          "a",
          "b",
          "c",
        ]
      }
      dataStoreName="dataStoreName"
      onCancel={[Function]}
      onStart={[Function]}
      selectedFeatures={
        Set {
          "b",
          "c",
          "a",
        }
      }
    />
  </div>
</Fragment>
`;
